src/sentry/tasks/statistical_detectors.py

```diff
@@ -544,13 +558,26 @@ def detect_function_trends(project_ids: List[int], start: datetime, *args, **kwa
     if not options.get("statistical_detectors.enable"):
         return
 
-    trends = _detect_function_trends(project_ids, start)
-    regressions = filter(lambda trend: trend[0] == TrendType.Regressed, trends)
+    ratelimit = options.get("statistical_detectors.ratelimit.ema")
+
+    regressions_by_project: DefaultDict[int, List[Tuple[float, DetectorPayload]]] = defaultdict(
+        list
+    )
+
+    for trend_type, score, payload in _detect_function_trends(project_ids, start):
+        if trend_type != TrendType.Regressed:
+            continue
+        heapq.heappush(regressions_by_project[payload.project_id], (score, payload))
+
+        while ratelimit >= 0 and len(regressions_by_project[payload.project_id]) > ratelimit:
+            heapq.heappop(regressions_by_project[payload.project_id])
```

Comment:
I think it would be better to have this be its own function/component that has clear inputs/outputs that can be tested in isolation.  that way you don't have to do all the mocking in your test to test it. I'll leave it up to you if you want to change that though

Reply:
I could pull this into a data structure that enforces the size limit but I would not remove the current test and its mocks because one important aspect of that test is that it's asserting the most significant regression is emitted not just a random one.

Reply:
I think a function would be sufficient

```python3
def rate_limit_regressions(rate_limit: int, detected_function_trends: Generator[Tuple[Optional[TrendType], float, DetectorPayload], None, None]) -> Generator[Tuple[Optional[TrendType], float, DetectorPayload], None, None]:
      regressions_by_project: DefaultDict[int, List[Tuple[float, DetectorPayload]]] = defaultdict(
        list
    )

    for trend_type, score, payload in _detect_function_trends(project_ids, start):
        if trend_type != TrendType.Regressed:
            continue
        heapq.heappush(regressions_by_project[payload.project_id], (score, payload))

        while ratelimit >= 0 and len(regressions_by_project[payload.project_id]) > ratelimit:
            heapq.heappop(regressions_by_project[payload.project_id])
    # emit the stuff that is not limited
```

then you can test just the rate limit function and make sure its logic behaves as you want. The mocking test is still good for understanding the integration but this way it's a separate component